File: solutions/cpp/863/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, TreeNode*> parents;
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_set<TreeNode *> seen;
        // First run a DFS to build the parents map
        dfs(root, nullptr);

        // Now run a BFS starting at the target node
        int steps = 0;
        queue<TreeNode *> queue;
        queue.push(target);
        seen.insert(target);

        vector<int> ans;
        while (queue.size() && steps < k) {
            int currLenght = queue.size();
            for (int i = 0; i < currLenght; i++) {
                TreeNode *node = queue.front();
                queue.pop();
                for (TreeNode *neighbor : {parents[node], node->left, node->right}) {
                    if (neighbor != nullptr && seen.find(neighbor) == seen.end()) {
                        queue.push(neighbor);
                        seen.insert(neighbor);
                    }
                }
            }
            steps++;
        }

        while (queue.size()) {
            TreeNode* node = queue.front();
            queue.pop();
            ans.push_back(node->val);
        }

        return ans;
    }

    void dfs(TreeNode *root, TreeNode *parent) {
        if (!root) {
            return;
        }
        parents[root] = parent;
        dfs(root->left, root);
        dfs(root->right, root);
    }
};

File: solutions/golang/863/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var parents map[*TreeNode]*TreeNode
var seen map[*TreeNode]bool

func distanceK(root *TreeNode, target *TreeNode, k int) []int {
	parents = make(map[*TreeNode]*TreeNode)
	seen = make(map[*TreeNode]bool)
	dfs(root, nil)

	queue := []*TreeNode{target}
	seen[target] = true
	steps := 0
	for len(queue) > 0 && steps < k {
		currLen := len(queue)
		for i := 0; i < currLen; i++ {
			node := queue[0]
			queue = queue[1:]
			neighbors := []*TreeNode{parents[node], node.Left, node.Right}
			for _, neighbor := range neighbors {
				_, ok := seen[neighbor]
				if neighbor != nil && !ok {
					seen[neighbor] = true
					queue = append(queue, neighbor)
				}
			}
		}
		steps++
	}

	var ans = []int{}
	for len(queue) > 0 {
		ans = append(ans, queue[0].Val)
		queue = queue[1:]
	}
	return ans
}

func dfs(root *TreeNode, parent *TreeNode) {
	if root == nil {
		return
	}

	parents[root] = parent
	dfs(root.Left, root)
	dfs(root.Right, root)
}